### PR TITLE
[llm] Use new API to register custom ops for llama model

### PR DESCRIPTION
**Pull request by larryliu0820**
Summary: Using the following 2 APIs:

* `EXECUTORCH_LIBRARY` to replace the need of a yaml file. With this macro we can directly register a custom kernel into ExecuTorch runtime.
* `WRAP_TO_ATEN` allows custom op authors to use the same kernel for ExecuTorch and PyTorch. This can be helpful during debugging.

Test Plan: Rely on the new CI job `test_llama` with `xnnpack+kv+custom` option.

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: backends/xnnpack/CMakeLists.txt [hunk 1]
```diff
@@ -72,6 +72,7 @@ target_include_directories(
   xnnpack_schema INTERFACE ${_xnnpack_schema__include_dir}
                            ${EXECUTORCH_ROOT}/third-party/flatbuffers/include)
 
+target_compile_options(pthreadpool PUBLIC ${_common_compile_options})
```
Comment (kimishpatel) [inline backends/xnnpack/CMakeLists.txt:75]: Why is this needed? and pthreadpool related stuff has moved to root CMakeLists

File: examples/models/llama2/CMakeLists.txt [hunk 2]
```diff
@@ -114,14 +128,13 @@ if(TARGET qnn_executorch_backend)
...
-# This one is needed for cpuinfo where it uses android
-# specific log lib
+# This one is needed for cpuinfo where it uses android specific log lib
 if(ANDROID)
   list(APPEND link_libraries log)
 endif()
 
 target_compile_options(llama_main PUBLIC ${_common_compile_options}
-  -DET_USE_THREADPOOL)
+                                         -DET_USE_THREADPOOL)
```
Comment (kimishpatel) [inline examples/models/llama2/CMakeLists.txt:137]: typo?

Comment (larryliu0820) [reply examples/models/llama2/CMakeLists.txt:137]: Reformat cmakelists.txt

File: backends/xnnpack/CMakeLists.txt (hunk 1 above)
Comment (larryliu0820) [reply backends/xnnpack/CMakeLists.txt:75]: my local build failed complaining no -fPIC

File: examples/models/llama2/custom_ops/op_sdpa_aot.cpp [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+    at::Tensor& value_cache,
+    const int64_t start_pos,
+    const int64_t seq_len,
+    const c10::optional<at::Tensor> attn_mask,
+    const double dropout_p,
+    const bool is_causal,
+    // @lint-ignore CLANGTIDY facebook-hte-ParameterMightThrowOnCopy
+    const c10::optional<double> scale) {
+  auto output = at::empty_like(q_projected);
+  WRAP_TO_ATEN(sdpa_with_kv_cache_out_no_context, 11)
```
Comment (kimishpatel) [inline examples/models/llama2/custom_ops/op_sdpa_aot.cpp:64]: whats this 11?

Comment (larryliu0820) [reply examples/models/llama2/custom_ops/op_sdpa_aot.cpp:64]: The 11th argument is `out`

Comment (kimishpatel) [reply examples/models/llama2/custom_ops/op_sdpa_aot.cpp:64]: # of args? I think there is some template magic that allows you to count number of args, right?

File: examples/models/llama2/custom_ops/op_sdpa_aot.cpp [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+} // namespace executor
+} // namespace torch
+
+TORCH_LIBRARY(llama, m) {
+  m.def(
+      "sdpa_with_kv_cache(Tensor query, Tensor key, Tensor value, Tensor(a!) key_cache, "
+      "Tensor(b!) value_cache, SymInt start_pos, SymInt seq_len, Tensor? attn_mask=None, "
+      "float drpout_p=0.0, bool is_causal=False, float? scale=None) -> Tensor",
+      &torch::executor::native::sdpa_with_kv_cache_aten);
+  m.def(
```
Comment (kimishpatel) [inline examples/models/llama2/custom_ops/op_sdpa_aot.cpp:90]: WHy is this one needed? So that we can generate out variant one?

File: examples/models/llama2/custom_ops/op_sdpa_with_kv_cache_test.cpp [hunk 5]
```diff
@@ -32,7 +32,7 @@ exec_aten::Tensor op_sdpa_with_kv_cache(
     exec_aten::optional<double> scale,
     exec_aten::Tensor& out) {
   exec_aten::RuntimeContext context{};
-  return torch::executor::llama::sdpa_with_kv_cache_outf(
+  return torch::executor::native::sdpa_with_kv_cache_out(
```
Comment (kimishpatel) [inline examples/models/llama2/custom_ops/op_sdpa_with_kv_cache_test.cpp:35]: why changes the namespace?

File: examples/models/llama2/custom_ops/sdpa_with_kv_cache.py [hunk 6]
```diff
@@ -4,196 +4,21 @@
...
+try:
+    op = torch.ops.llama.sdpa_with_kv_cache.default
+    assert op is not None
+except:
+    # assuming we only hit this in OSS, find the default install path
+    prefix = os.environ.get("CMAKE_INSTALL_PREFIX", "../../../../cmake-out")
+    lib_path = os.path.join(prefix, "lib/libcustom_ops_aot_lib.so")
+    torch.ops.load_library(lib_path)
+    op = torch.ops.llama.sdpa_with_kv_cache.default
+    assert op is not None
```
Comment (kimishpatel) [inline examples/models/llama2/custom_ops/sdpa_with_kv_cache.py:24]: This is a little bit clunky but I dont know how to improv e it

File: examples/models/llama2/custom_ops/targets.bzl [hunk 7]
```diff
@@ -2,103 +2,23 @@ load("@fbsource//xplat/executorch/build:runtime_wrapper.bzl", "runtime")
...
-        define_static_targets = True,
-        visibility = [
-            "//executorch/codegen/...",
-            "@EXECUTORCH_CLIENTS",
-        ],
-    )
-
     runtime.cxx_library(
-        name = "sdpa",
+        name = "custom_ops",
```
Comment (kimishpatel) [inline examples/models/llama2/custom_ops/targets.bzl:13]: why rename this?

**Review by kimishpatel — CHANGES_REQUESTED** [review]
Left some comments

File: examples/models/llama2/custom_ops/op_sdpa_with_kv_cache_test.cpp (hunk 5 above)
Comment (larryliu0820) [reply examples/models/llama2/custom_ops/op_sdpa_with_kv_cache_test.cpp:35]: `llama` namespace was generated by `FunctionHeaderWrapper.h`. Here I added a header `op_sdpa.h` and that is using the same namespace as `op_sdpa.cpp`.

File: examples/models/llama2/custom_ops/targets.bzl (hunk 7 above)
Comment (larryliu0820) [reply examples/models/llama2/custom_ops/targets.bzl:13]: A lot of places are referring to `examples/models/llama2/custom_ops:custom_ops` library name. I'm just too lazy to change all of them to `sdpa`.

File: examples/models/llama2/custom_ops/op_sdpa_aot.cpp (hunk 3 above)
Comment (larryliu0820) [reply examples/models/llama2/custom_ops/op_sdpa_aot.cpp:64]: yeah this is telling the template we need to return the 11th object, since it is `out`. See this code: https://github.com/pytorch/executorch/blob/main/extension/aten_util/make_aten_functor_from_et_functor.h#L268-L277

File: examples/models/llama2/custom_ops/op_sdpa_aot.cpp (hunk 4 above)
Comment (larryliu0820) [reply examples/models/llama2/custom_ops/op_sdpa_aot.cpp:90]: We need both `sdpa_with_kv_cache` and `sdpa_with_kv_cache.out` in ATen so that exir is happy.

**Review by kimishpatel — APPROVED** [review]
